packages/commonwealth/server/migrations/20230217164706-root-id-refactor.js

```diff
@@ -0,0 +1,134 @@
+'use strict';
+
+// The purpose of this script is to turn root_id to thread_id. To achieve this without losing any data, we must set the
+// comments that pointed to a proposal, to instead point to a thread that is created to backlink to that proposal
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    await queryInterface.sequelize.transaction(async (t) => {
+      const updateQueries = [];
+
+      const threadComments = await queryInterface.sequelize.query(
+        `SELECT * FROM "Comments" WHERE "root_id" ILIKE '%discussion%'`
+      );
+
+      // for each comment that is part of a discussion, replace its root_id with the thread_id
+      threadComments[0].forEach((c) => {
+        updateQueries.push(
+          queryInterface.sequelize.query(
+            `UPDATE "Comments" SET "root_id" = ${
+              c.root_id.split('_')[1]
+            } WHERE "id"=${c.id}`,
+            { transaction: t }
+          )
+        );
+      });
+
+      // we will now perform the same operation with proposal comments, but in order to do so we must create threads
+      // to backlink to proposals
+      const proposalComments = await queryInterface.sequelize.query(
+        `SELECT * FROM "Comments" WHERE "root_id" NOT ILIKE '%discussion%'`
+      );
+
+      // maps root_id -> newThread it creates
+      const newThreads = new Map();
+      proposalComments[0].forEach((c) => {
+        if (newThreads.get(c.root_id)) {
+          return; //if we already had created thread to backlink proposal comments, don't create another one
+        }
+
+        let rootParts = c.root_id.split('_');
+        // handle cosmosproposals differently
+        if (rootParts[0] === 'cosmosproposal') {
+          rootParts = ['proposal', rootParts[1]];
+        } else {
+          rootParts = [`proposal/${rootParts[0]}`, rootParts[1]];
+        }
+
+        // otherwise map the root_id to a new thread to backlink proposal comments
+        newThreads.set(c.root_id, {
+          address_id: 1,
+          title: `Thread for ${c.chain}'s ${rootParts[0]} ${rootParts[1]}`,
+          kind: 'link',
+          body: '',
+          url: `https://commonwealth.im/${c.chain}/${rootParts[0]}/${rootParts[1]}`,
+          chain: c.chain,
+          read_only: true,
+          stage: 'discussion',
+          canvas_action: null,
+          canvas_hash: null,
+          canvas_session: null,
+          // set the dates to a year back so they dont spam chain's home page
+          created_at: new Date(
+            new Date().setFullYear(new Date().getFullYear() - 1)
+          ),
+          updated_at: new Date(
+            new Date().setFullYear(new Date().getFullYear() - 1)
+          ),
+        });
+      });
+
+      const inserted = await queryInterface.bulkInsert(
+        'Threads',
+        [...newThreads.values()],
+        { returning: true, transaction: t }
+      );
+
+      // create map of rootId -> thread_id
```

Comment:
Can we construct this map as part of the thread creation loop, so we don't need to parse out the url?

Reply:
Done, but we still need urls because only threads get returned from the bulkInsert, so as a result we need some unique  key that contains both chain and proposal information. So url is the best in this case since it has both.